src/realm/sync/noinst/client_history_impl.cpp

```diff
@@ -359,128 +359,161 @@ void ClientHistory::integrate_server_changesets(const SyncProgress& progress,
+                    ++changeset_ndx;
+                }
+
+                subsets_to_integrate = subsets_to_integrate - 1;
+
+                if (m_db->waiting_for_write_lock()) {
+                    // Stop if a user is waiting to commit.
+                    break;
+                }
+            }
+        }
+        catch (BadChangesetError& e) {
+            throw IntegrationException(ClientError::bad_changeset,
+                                       util::format("Failed to apply received changeset: %1", e.what()));
+        }
+        catch (TransformError& e) {
+            throw IntegrationException(ClientError::bad_changeset,
+                                       util::format("Failed to transform received changeset: %1", e.what()));
+        }
 
-    version_type new_version = transact->commit_and_continue_as_read().version; // Throws
+        // downloaded_bytes always contains the total number of downloaded bytes
+        // from the Realm. downloaded_bytes must be persisted in the Realm, since
+        // the downloaded changesets are trimmed after use, and since it would be
+        // expensive to traverse the entire history.
+        Array& root = m_arrays->root;
+        auto downloaded_bytes =
+            std::uint_fast64_t(root.get_as_ref_or_tagged(s_progress_downloaded_bytes_iip).get_as_int());
+        downloaded_bytes += downloaded_bytes_in_subset;
+        root.set(s_progress_downloaded_bytes_iip, RefOrTagged::make_tagged(downloaded_bytes)); // Throws
+
+        // The reason we can use the `origin_timestamp`, and the `origin_file_ident`
+        // from the last changeset in the subset, and ignore all the other changesets, is
+        // that these values are actually irrelevant for changesets of remote origin
+        // stored in the client-side history (for now), except that
+        // `origin_file_ident` is required to be nonzero, to mark it as having been
+        // received from the server.
+        const Changeset& last_changeset = changesets[changeset_ndx - 1];
+        HistoryEntry entry;
+        entry.origin_timestamp = last_changeset.origin_timestamp;
+        entry.origin_file_ident = last_changeset.origin_file_ident;
+        entry.remote_version = last_changeset.version;
+
+        // m_changeset_from_server is picked up by prepare_changeset(), which then
+        // calls add_sync_history_entry(). prepare_changeset() is called as a result
+        // of committing the current transaction even in the "short-circuited" mode,
+        // because replication isn't disabled.
+        REALM_ASSERT(!m_changeset_from_server);
+        m_changeset_from_server = entry;
+
+        // During the bootstrap phase in flexible sync, the server sends multiple download messages with the same
+        // synthetic server version that represents synthetic changesets generated from state on the server.
+        if (changeset_ndx == num_changesets && batch_state == DownloadBatchState::LastInBatch) {
```

Comment:
Is this right? I think we actually want to update the progress so that the download cursor reflects the last changeset we integrated from the server so that new local changesets can have that as the changeset they're based on here https://github.com/realm/realm-core/blob/17a48eed2e93661ef4efca50ff26de30fe1ae3db/src/realm/sync/noinst/client_history_impl.cpp#L658

Reply:
great catch!